QUARKUS-3456.md

```diff
@@ -0,0 +1,65 @@
+
+Upstream PR: https://github.com/quarkusio/quarkus/pull/34493
+
+Red Hat Build of Quarkus has ability to handle secrets since the Ghost release.
+This feature worked alright, but encrypting secrets so that built-in secret handlers can decode them was somehow awkward.
+It was far from trivial to find a correct steps and encrypt the secret in a way Quarkus accepted.
+Now Quarkus leverages Quarkus CLI and allows to create, update and remove configuration properties with it.
+One of options is to pass the CLI plain secret and Quarkus will encode this secret for you.
+You can even ask Quarkus CLI to set encoded secret directly to the application properties.
+Or, if you use YAML or other formats (like if you keep secrets in the environment variables), you can just copy it from the CLI output.
+We should test this new feature, including other sub-commands under `quarkus config` command as users are likely to use them when they encrypt secrets.
+
+## Scope of the testing
+Test coverage must include:
+
+- `quarkus config set/remove/encrypt` command tested with options combined in a fashion customers are likely to use,
+- documented flow to create a secret, add it to the Keystore, set Keystore config source and check Quarks application decodes secrets,
+- default secret handler `AES/GCM/NoPadding` is used to create the secret,
+- secret is decoded in FIPS-enabled environment,
+- read encrypted configuration in Quarkus application user code,
+- store secrets generated with the CLI in a Keystore and access them using Quarkus application,
+- decode the encryption key in Base64, if the plain text key was used to encrypt the secret.
+
+## Getting familiar with the feature
+Following actions were taken to ensure familiarity:
+- Focus on exploratory testing of the feature
+- Ensure good user experience and simplicity of use
+
+## Existing test coverage
+Upstream test coverage in Quarkus project includes test of encryption with a plain encoded secret, plain secret.
+Also, encryption tests with explicitly declared key and with a generated key.
+Most importantly, the PR closed [Quarkus issue #29172](https://github.com/quarkusio/quarkus/issues/29172), however it does test it partially and all the capabilities are tested in isolation.
+
+We test secret key handlers in the `config` module of the Quarkus QE Test Suite.
+
+## Impact on test suites and testing automation
+Three new integration test classes will be added to Quarkus CLI module of the Quarkus QE Test Suite.
+One for each `quarkus config` subcommand (`set`, `remove`, `encrypt`).
+Tests will run in a JVM mode only as intention is to test CLI and assure that `application.properties` are updated correctly. 
+
+## Impact on resources
+We will start JVM application once per each test class and call an endpoint or two, therefore additional test execution time should be below 30 seconds.
+
+## Future considerations
+
+- use more Ciphers (use different algorithms, modes and padding) to encrypt a secret
```

Comment:
I would prefer to move it out of the future scope and put it into current TD.
What would be the estimate to implement this?

Reply:
You can chose alg/mode/padding via CLI encrypt cmd options https://github.com/quarkusio/quarkus/blob/main/devtools/cli/src/main/java/io/quarkus/cli/config/Encrypt.java#L59 but I am worried about this https://github.com/quarkusio/quarkus/blob/main/devtools/cli/src/main/java/io/quarkus/cli/config/Encrypt.java#L64. I simply didn't get to test this and I don't know if you can use different cipher while you still have there GCM spec 128 and sha256.

I cannot tell if it is additional time or not, but https://github.com/quarkusio/quarkus/pull/34493 said `Can encrypt arbitrary values in AES/GCM/NoPadding (will support more algorithms)` so I didn't expect it is actually supported. Right now, it is not documented which other ciphers are supported.

Reply:
@rsvoboda 
```
    @Option(hidden = true, names = { "-a", "--algorithm" }, description = "Algorithm", defaultValue = "AES")
    String algorithm;

    @Option(hidden = true, names = { "-m", "--mode" }, description = "Mode", defaultValue = "GCM")
    String mode;

    @Option(hidden = true, names = { "-p", "--padding" }, description = "Padding", defaultValue = "NoPadding")
    String padding;

    @Option(hidden = true, names = { "-q", "--quiet" }, defaultValue = "false")
    boolean quiet;
```

they are hidden, that and the fact it is not documented says it all, it is not supported ATM IMHO

Reply:
https://issues.redhat.com/browse/QUARKUS-3456?focusedId=25161121&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-25161121